### PR TITLE
feat: editing townsquare comments

### PR DESCRIPTION
**Pull request by walidmujahid**
Editing a comment is as simple as clicking the edit button on a comment. All other comments of the user will have their edit buttons disabled when editing of a comment is active. Posting new comments for the user will also be disabled while they are editing a comment. If you try to submit an edit with zero characters, it will not go through. An edited comment will show `(edited)` string next to the `timAgo` section.

**2020-02-30 - Working Demo - Video:** https://youtu.be/wibE0emPwBE
**2020-02-22 - Video:** https://youtu.be/H9tvgp3i0y8

- [x] feat: editing townsquare comments
    - [x] add edit button and handler function
    - [x] handle an `EDIT` method in api
    - [x] add an 'edited' tag to show the comment was edited
- [x] fix introduced bugs
    - [x] fix issue with empty comment container being disabled after hitting 'Enter'
- [x] code review
- [x] refactor code 
- [x] maintainer review

closes https://github.com/gitcoinco/web/issues/6010

### CODE REVIEW

**Comment by frankchen07** [comment]
@walidmujahid - this looks pretty dope! thank you for the videos!

1. I didn't see if the trash can (delete comment) functionality works as intended, but I'm sure it does.

2. wanted to make sure that only the user who made the comment can edit it (I think this was obvious but perhaps hard to show in the video).

**Comment by walidmujahid** [comment]
@frankchen07 Deleting comments is actually feature that @thelostone-mc -if I am not mistaken- had rolled out. 

> wanted to make sure that only the user who made the comment can edit it (I think this was obvious but perhaps hard to show in the video).

Yes, only the user who posted the comment can edit. I actually tested this with the very useful user "impersonation" tool, but I did not show it in any video.

File: app/assets/v2/js/activity.js [hunk 1]
```diff
@@ -303,53 +303,95 @@ $(document).ready(function() {
...
-    // remote post
-    var params = {
-      'method': 'comment',
-      'comment': comment,
-      'csrfmiddlewaretoken': $('input[name=csrfmiddlewaretoken]').val()
-    };
-    var url = '/api/v0.1/activity/' + $parent.data('pk');
+      let comment_id = $('.editableComment').parent().closest('.comment_row').data('id');
+      let url = '/api/v0.1/comment/' + comment_id;
+      let params = {
```
Comment (thelostone-mc) [inline app/assets/v2/js/activity.js]: let -> const

File: app/assets/v2/js/activity.js [hunk 2]
```diff
@@ -303,53 +303,95 @@ $(document).ready(function() {
...
 
-    // remote post
-    var params = {
-      'method': 'comment',
-      'comment': comment,
-      'csrfmiddlewaretoken': $('input[name=csrfmiddlewaretoken]').val()
-    };
-    var url = '/api/v0.1/activity/' + $parent.data('pk');
+      let comment_id = $('.editableComment').parent().closest('.comment_row').data('id');
+      let url = '/api/v0.1/comment/' + comment_id;
```
Comment (thelostone-mc) [inline app/assets/v2/js/activity.js]: let -> const

File: app/assets/v2/js/activity.js [hunk 3]
```diff
@@ -303,53 +303,95 @@ $(document).ready(function() {
...
+      $parent.parents('.activity.box').find('.loading').removeClass('hidden');
 
-    // remote post
-    var params = {
-      'method': 'comment',
-      'comment': comment,
-      'csrfmiddlewaretoken': $('input[name=csrfmiddlewaretoken]').val()
-    };
-    var url = '/api/v0.1/activity/' + $parent.data('pk');
+      let comment_id = $('.editableComment').parent().closest('.comment_row').data('id');
```
Comment (thelostone-mc) [inline app/assets/v2/js/activity.js]: let -> const

File: app/assets/v2/js/activity.js [hunk 4]
```diff
@@ -303,53 +303,95 @@ $(document).ready(function() {
...
+      };
 
-    $.post(url, params, function(response) {
-      var success_callback = function($parent) {
-        $parent.parents('.box').find('.comment_container textarea').val('');
-        $parent.find('textarea').focus();
+      $.post(url, params, function(response) {
+        let success_callback = function($parent) {
+          // $parent.find('textarea').focus();
+        };
```
Comment (thelostone-mc) [inline app/assets/v2/js/activity.js]: should we remove this if it's not doing anything?

File: app/assets/v2/js/activity.js [hunk 5]
```diff
@@ -303,53 +303,95 @@ $(document).ready(function() {
...
+      };
+      var url = '/api/v0.1/activity/' + $parent.data('pk');
+
+      $.post(url, params, function(response) {
+        var success_callback = function($parent) {
+          $parent.parents('.box').find('.comment_container textarea').val('');
+          $parent.find('textarea').focus();
+
+        };
+        var override_hide_comments = !has_hidden_comments;
```
Comment (thelostone-mc) [inline app/assets/v2/js/activity.js]: `const`

File: app/assets/v2/js/activity.js [hunk 6]
```diff
@@ -303,53 +303,95 @@ $(document).ready(function() {
...
+      num = parseInt(num) + 1;
+      $parent.find('span.num').html(num);
+
+      // remote post
+      var params = {
+        'method': 'comment',
+        'comment': comment,
+        'csrfmiddlewaretoken': $('input[name=csrfmiddlewaretoken]').val()
+      };
+      var url = '/api/v0.1/activity/' + $parent.data('pk');
```
Comment (thelostone-mc) [inline app/assets/v2/js/activity.js]: `const`

File: app/assets/v2/js/activity.js [hunk 7]
```diff
@@ -303,53 +303,95 @@ $(document).ready(function() {
...
+      $parent.parents('.activity.box').find('.loading').removeClass('hidden');
+      var has_hidden_comments = $parent.parents('.activity.box').find('.row.comment_row.hidden').length;
+      // increment number
+      var num = $parent.find('span.num').html();
+
+      num = parseInt(num) + 1;
+      $parent.find('span.num').html(num);
+
+      // remote post
+      var params = {
```
Comment (thelostone-mc) [inline app/assets/v2/js/activity.js]: `const`

File: app/assets/v2/js/activity.js [hunk 8]
```diff
@@ -303,53 +303,95 @@ $(document).ready(function() {
...
+
+      // validation
+      if (!comment) {
+        return;
+      }
+
+      $parent.parents('.activity.box').find('.loading').removeClass('hidden');
+      var has_hidden_comments = $parent.parents('.activity.box').find('.row.comment_row.hidden').length;
+      // increment number
+      var num = $parent.find('span.num').html();
```
Comment (thelostone-mc) [inline app/assets/v2/js/activity.js]: `const`

File: app/assets/v2/js/activity.js [hunk 9]
```diff
@@ -303,53 +303,95 @@ $(document).ready(function() {
...
+      $parent.parents('.box').find('.comment_container textarea').prop('disabled', true);
+      $('.post_comment').prop('disabled', true);
+
+      // validation
+      if (!comment) {
+        return;
+      }
+
+      $parent.parents('.activity.box').find('.loading').removeClass('hidden');
+      var has_hidden_comments = $parent.parents('.activity.box').find('.row.comment_row.hidden').length;
```
Comment (thelostone-mc) [inline app/assets/v2/js/activity.js]: `const`

File: app/assets/v2/js/activity.js [hunk 10]
```diff
@@ -303,53 +303,95 @@ $(document).ready(function() {
...
+        $(`.comment_row[data-id="${comment_id}"] .comment_options`).prop('disabled', false).removeClass('hidden');
+        $(`.comment_row[data-id="${comment_id}"]`).parent().closest('.box')
+          .find('#container-post_comment')
+          .find('*').prop('disabled', false).removeClass('hidden');
+        // TODO: automatically cancel editing when another edit_comment button is clicked
+        $('.edit_comment').prop('disabled', false);
       });
+    } else {
+      // user input
+      var comment = $parent.parents('.box').find('.comment_container textarea').val();
```
Comment (thelostone-mc) [inline app/assets/v2/js/activity.js]: `const`

File: app/assets/v2/js/activity.js [hunk 11]
```diff
@@ -303,53 +303,95 @@ $(document).ready(function() {
...
+      // remote post
+      var params = {
+        'method': 'comment',
+        'comment': comment,
+        'csrfmiddlewaretoken': $('input[name=csrfmiddlewaretoken]').val()
+      };
+      var url = '/api/v0.1/activity/' + $parent.data('pk');
+
+      $.post(url, params, function(response) {
+        var success_callback = function($parent) {
```
Comment (thelostone-mc) [inline app/assets/v2/js/activity.js]: let / const

File: app/assets/v2/js/activity.js [hunk 12]
```diff
@@ -303,53 +303,95 @@ $(document).ready(function() {
...
+        // pass
+      })
+        .fail(function() {
+          $parent.parents('.activity.box').find('.error').removeClass('hidden');
+        })
+        .always(function() {
+          $parent.parents('.activity.box').find('.loading').addClass('hidden');
+          $parent.parents('.box').find('.comment_container textarea').prop('disabled', false);
+          $('.post_comment').prop('disabled', false);
+        });
```
Comment (thelostone-mc) [inline app/assets/v2/js/activity.js]: ^de-indent by 1

File: app/assets/v2/js/activity.js [hunk 13]
```diff
@@ -581,6 +628,59 @@ $(document).ready(function() {
...
+    });
+  });
+
+
+  $(document).on('click', '.edit_comment', function(e) {
+    e.preventDefault();
+    let comment_id = $(this).data('pk');
+    let commentContainer = $(`.comment_row[data-id="${comment_id}"] .activity_comments_main_comment`);
+    let content = commentContainer.html().replace(/<br>/g, '\n').trim();
+    let editableContainer = $(`<div id="editableContainer" class="col-12 col-sm-12 text-right"><textarea class="form-control bg-lightblue font-caption editableComment" cols="80" rows="3" data-id="${comment_id}">${content}</textarea><a href=# class="btn btn-gc-blue btn-sm mt-2 font-smaller-7 font-weight-bold cancel_edit" data-id="${comment_id}">CANCEL</a></div>`);
```
Comment (thelostone-mc) [inline app/assets/v2/js/activity.js]: ^ update all the `let` to `const` if it's not being over-written ?

File: app/assets/v2/js/activity.js [hunk 14]
```diff
@@ -608,12 +708,42 @@ $(document).ready(function() {
...
+      let comment_id = $(this).data('id');
+      let editableContainer = $('#editableContainer');
+
+      let url = '/api/v0.1/comment/' + comment_id;
+      let params = {
+        'method': 'GET_COMMENT'
+      };
+
+      $.get(url, params, function(response) {
+        let the_comment = response['comment'];
```
Comment (thelostone-mc) [inline app/assets/v2/js/activity.js]: ^ update all the let to const if it's not being over-written ?

File: app/assets/v2/js/activity.js (hunk 4 above)
Comment (walidmujahid) [reply app/assets/v2/js/activity.js]: The commented line or the entire `success_callback` block @thelostone-mc?

File: app/assets/v2/js/activity.js (hunk 12 above)
Comment (walidmujahid) [reply app/assets/v2/js/activity.js]: Are you sure? Travis's style check suggested that be indented in the first place.

Comment (walidmujahid) [reply app/assets/v2/js/activity.js]: de-indented

File: app/assets/v2/js/activity.js (hunk 4 above)
Comment (walidmujahid) [reply app/assets/v2/js/activity.js]: replaced commented out code with `$.noop(); // do nothing`

**Comment by walidmujahid** [comment]
@thelostone-mc I have switched most of the `let` to `const` where applicable. I did deindent on your suggestion, but as I mentioned before, eslint suggest otherwise, so travis gives a style error for that currently.

I also tested everything again after merging the updates from master and everything seems to work. There is, however, a small issue that I am not _yet_ sure if my code is introducing or if it is from master. The comment count on an activity shows as zero and the comments view seems automatically closed.

I will debug and update.

**Comment by walidmujahid** [comment]
**Update:**

Okay, the issue is not with in this PR but was introduced in a commit -not sure which yet- that is in master.

This feature works as expected. I have made the requested changes @thelostone-mc

**Comment by walidmujahid** [comment]
This is ready to be reviewed again @thelostone-mc, @octavioamu, @danlipert

**Comment by walidmujahid** [comment]
I merged with the recent upstream and I have yet to test it. I have been having some local problems, possibly with UTF8 encodings that may be affecting my builds in other areas than just described here: https://github.com/gitcoinco/web/issues/6136

I will try to figure this out in the morning, EST, when I wake up.

**Comment by walidmujahid** [comment]
I finished testing everything on my end after merging with recent updates. I also merged with my fork's master branch to include the .gitattributes from: https://github.com/gitcoinco/web/pull/6175

This is ready for any further review @octavioamu @thelostone-mc @danlipert

**Comment by walidmujahid** [comment]
I have finished merging the most recent upstream changes and I have locally tested the feature. Everything works as excepted and shown in the demo videos.

**Comment by walidmujahid** [comment]
Merged recent upstream changes and tested locally. The feature works as expected. 

@owocki @thelostone-mc @danlipert @octavioamu This is ready for review.

**Comment by walidmujahid** [comment]
@octavioamu I update this soon and deal with any merge conflicts on my end.

**Comment by octavioamu** [comment]
> @octavioamu I will update this soon and deal with any merge conflicts on my end.

I use to move back to an older migration before these, then manually remove them and run makemigrations again .

**Comment by walidmujahid** [comment]
> I use to move back to an older migration before these, then manually remove them and run makemigrations again .

@octavioamu  I just merge the conflicting migrations, but should I delete migrations to the current upstream head and re-run makemigrations like you do?

**Comment by walidmujahid** [comment]
Sorry for my confusing replies. I just reread your messages and noticed 'requested changes'. I will have this updated soon.

**Comment by walidmujahid** [comment]
@octavioamu I have worked on the migrations issue. I hope it is better now :-)

**Comment by thelostone-mc** [comment]
@walidmujahid so you migrations you've deleted ,  a few would aleady been in live
If we merge this in -> things would get messy
Could you simply readd all the migrations files which you haven't created 
And then 
ensure there is a single migration file for all your model changes

**Comment by walidmujahid** [comment]
Yes. Sorry about that. I will do that as soon as possible.

**Comment by walidmujahid** [comment]
> @walidmujahid so you migrations you've deleted , a few would aleady been in live
> If we merge this in -> things would get messy
> Could you simply readd all the migrations files which you haven't created
> And then
> ensure there is a single migration file for all your model changes

Resolved @thelostone-mc

**Comment by walidmujahid** [comment]
@danlipert  @thelostone-mc @octavioamu I have updated this with most recent upstream changes and have recreated the migrations. Manual smoke testing for comment editing functionality passed.

**Comment by walidmujahid** [comment]
@danlipert  @thelostone-mc @octavioamu I have updated this with most recent upstream master changes and have recreated the migrations. Manual smoke testing for comment editing functionality passed. Is there anything preventing this from merged into master?

**Comment by thelostone-mc** [comment]
@walidmujahid nah it looks good but looks like CI fails due to eslint errors !
It's indentation fix from the looks of it -> if you can get those in, we can try and get this live this week

**Comment by thelostone-mc** [comment]
@walidmujahid nvm I fixed it up !

**Comment by walidmujahid** [comment]
> @walidmujahid nvm I fixed it up !

Oh, okay. You were quick :-)

File: app/townsquare/views.py [hunk 15]
```diff
@@ -472,6 +472,8 @@ def api(request, activity_id):
             counter += 1; results[counter] += time.time() - start_time; start_time = time.time()
             comment_dict['sorted_match_curve'] = comment.profile.matchranking_this_round.sorted_match_curve if comment.profile.matchranking_this_round else None
             counter += 1; results[counter] += time.time() - start_time; start_time = time.time()
+            if comment.is_edited:
```
Comment (octavioamu) [inline app/townsquare/views.py:475]: is this in any point making a validation to know if the user editing the comment is actually the comment owner? worry about people editing others comments

Comment (thelostone-mc) [reply app/townsquare/views.py:475]: It's simply to add a label right ? Should be alright

Comment (walidmujahid) [reply app/townsquare/views.py:475]: @octavioamu Actually, I completely missed this. This not actually validating things. I will add that quickly.

Comment (walidmujahid) [reply app/townsquare/views.py:475]: Actually, never mind, @octavioamu. @thelostone-mc had written the validation intially for the `comment_v1` views endpoint at: 

https://github.com/gitcoinco/web/blob/3b0b1e1a3daf07739c4e683d781c5e9e11ee8405/app/townsquare/views.py#L589-L594

So the comment.is_edited is for the label as @thelostone-mc said @octavioamu. The endpoint has validation to make sure a user does not edit except their own comments.

**Pull request merged**